pkg/solana/chainreader/event_read_binding.go

```diff
@@ -0,0 +1,229 @@
+
+	logs, err := b.events.FilteredLogs(ctx, filter.Expressions, limitAndSort, itemType)
+	if err != nil {
+		return nil, err
+	}
+
+	sequences, err := b.decodeLogsIntoSequences(ctx, logs, sequenceDataType)
+	if err != nil {
+		return nil, err
+	}
+
+	return sequences, nil
+}
+
+func (b *eventReadBinding) remapPrimitive(expression query.Expression) (query.Expression, error) {
+	var (
+		comp query.Expression
+		err  error
+	)
+
+	switch primitive := expression.Primitive.(type) {
+	case *primitives.Comparator:
+		if comp, err = b.encodeComparator(primitive); err != nil {
+			return query.Expression{}, fmt.Errorf("failed to encode comparator %q: %w", primitive.Name, err)
+		}
+	case *primitives.Confidence:
+		// TODO: maybe use an ignore filter?
+		// confidence is ignored in solana
+	}
+
+	return comp, nil
+}
+
+func (b *eventReadBinding) encodeComparator(comparator *primitives.Comparator) (query.Expression, error) {
+	onChainName, ok := b.indexedSubkeys[comparator.Name]
+	if !ok {
+		return query.Expression{}, fmt.Errorf("%w: unknown indexed subkey mapping %s", types.ErrInvalidConfig, comparator.Name)
+	}
+
+	itemType := strings.Join([]string{b.namespace, b.genericName, comparator.Name}, ".")
+
+	for idx, comp := range comparator.ValueComparators {
+		// need to do a transform and then extract the value for the subkey
+		newValue, err := b.modifier.TransformToOnChain(comp.Value, itemType)
+		if err != nil {
+			return query.Expression{}, err
+		}
+
+		comparator.ValueComparators[idx].Value = newValue
+	}
+
+	return logpoller.NewEventSubkeyFilter(strings.Split(onChainName, "."), comparator.ValueComparators), nil
+}
+
+func (b *eventReadBinding) decodeLogsIntoSequences(
+	ctx context.Context,
+	logs []logpoller.Log,
+	into any,
+) ([]types.Sequence, error) {
+	sequences := make([]types.Sequence, len(logs))
+
+	for idx := range logs {
+		sequences[idx] = types.Sequence{
+			Cursor: logpoller.FormatContractReaderCursor(logs[idx]),
+			Head: types.Head{
+				Height:    fmt.Sprint(logs[idx].BlockNumber),
+				Hash:      solana.PublicKey(logs[idx].BlockHash).Bytes(),
+				Timestamp: uint64(logs[idx].BlockTimestamp.Unix()), //nolint:gosec
+			},
+		}
+
+		var typeVal reflect.Value
+
+		typeInto := reflect.TypeOf(into)
+		if typeInto.Kind() == reflect.Pointer {
+			typeVal = reflect.New(typeInto.Elem())
+		} else {
+			typeVal = reflect.Indirect(reflect.New(typeInto))
+		}
+
+		// create a new value of the same type as 'into' for the data to be extracted to
+		sequences[idx].Data = typeVal.Interface()
+
+		if err := b.decodeLog(ctx, &logs[idx], sequences[idx].Data); err != nil {
+			return nil, err
+		}
+	}
+
+	return sequences, nil
+}
+
+func (b *eventReadBinding) decodeLog(ctx context.Context, log *logpoller.Log, into any) error {
+	itemType := codec.WrapItemType(false, b.namespace, b.genericName, codec.ChainConfigTypeEventDef)
+
+	// decode non indexed topics and apply output modifiers
```

Comment:
This is an evm-specific comment. For Solana, it should just say "decode event data".

Reply:
Good callout.